### PR TITLE
Support S3 data loading

### PR DESCRIPTION
**Pull request by jrocmar**
This PR introduces ~~the S3IndexedDataset, which supports loading a dataset stored in S3 in the same format as the MMapIndexedDataset~~ S3 data loading to IndexedDataset. In particular, the .idx file is downloaded to a local directory at initialization so that we can memory map it and the .bin file is streamed into memory block-by-block.

### CODE REVIEW

**Comment by jrocmar** [comment]
Draft implementation for #698

File: megatron/core/datasets/s3_indexed_dataset.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+    assert _S3_PREFIX + bucket + "/" + key == path
+    return _ParsedS3Path(bucket=bucket, key=key)
+
+
+def _object_exists(s3_client, path: str) -> bool:
+    parsed_s3_path = _parse_s3_path(path)
+    try:
+        response = s3_client.head_object(Bucket=parsed_s3_path.bucket, Key=parsed_s3_path.key)
+        return True
+    except ClientError:
```
Comment (jkamalu) [inline megatron/core/datasets/s3_indexed_dataset.py]: A ClientError will be thrown in the case of any failure correct, even in the object exists? For example, we should differentiate between bad permissions and a missing object.

File: megatron/core/datasets/s3_indexed_dataset.py [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+
+
+class _S3Agent:
+    """Retrieve byte ranges from S3 for the S3IndexedDataset."""
+
+    def __init__(self, path, cache_nbytes):
+        self._client = boto3.client('s3')
+        assert path.startswith('s3://')
+        path = path[len('s3://') :]
+        self._bucket, self._key = path.split('/', 1)
```
Comment (jkamalu) [inline megatron/core/datasets/s3_indexed_dataset.py]: Use globals

`_S3_PREFIX`

`_parse_s3_path`

File: megatron/core/datasets/s3_indexed_dataset.py [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+def _object_exists(s3_client, path: str) -> bool:
+    parsed_s3_path = _parse_s3_path(path)
+    try:
+        response = s3_client.head_object(Bucket=parsed_s3_path.bucket, Key=parsed_s3_path.key)
+        return True
+    except ClientError:
+        return False
+
+
+class _S3Agent:
```
Comment (jkamalu) [inline megatron/core/datasets/s3_indexed_dataset.py]: This class, though private, should have arg and return typing for all functions

File: megatron/core/datasets/s3_indexed_dataset.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+        )['Body'].read()
+        self._cache_bytes_start = bytes_start
+        self._cache_bytes_end = bytes_end
+        return self._extract_from_cache(offset, size)
+
+    def close(self):
+        self._client.close()
+
+
+def _download_file(s3_client, s3_path, local_path):
```
Comment (jkamalu) [inline megatron/core/datasets/s3_indexed_dataset.py]: Needs typing and doc string

File: megatron/core/datasets/s3_indexed_dataset.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+    dirname = os.path.dirname(local_path)
+    os.makedirs(dirname, exist_ok=True)
+    parsed_s3_path = _parse_s3_path(s3_path)
+    s3_client.download_file(parsed_s3_path.bucket, parsed_s3_path.key, local_path)
+
+
+def _maybe_download_file(s3_path, local_path):
```
Comment (jkamalu) [inline megatron/core/datasets/s3_indexed_dataset.py]: Needs typing and doc string, in particular a description of when maybe means no and when it means yes

File: megatron/core/datasets/s3_indexed_dataset.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+    if torch.distributed.is_initialized():
+        rank = torch.distributed.get_rank()
+        # https://github.com/NVIDIA/Megatron-LM/blob/\
+        # 89574689447d694bb19dd86fc8a6153b4467ba9d/megatron/initialize.py#L232
```
Comment (jkamalu) [inline megatron/core/datasets/s3_indexed_dataset.py]: What is this for?

File: megatron/core/datasets/s3_indexed_dataset.py [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+        _download_file(s3_client, s3_path, local_path)
+
+    if torch.distributed.is_initialized():
+        torch.distributed.barrier()
+
+    # If the `local_path` is in a file system that is not
+    # shared across all the ranks, then we assume it's in the
+    # host file system and each host needs to download the file.
+    if (not os.path.exists(local_path)) and (local_rank == 0):
+        _download_file(s3_client, s3_path, local_path)
```
Comment (jkamalu) [inline megatron/core/datasets/s3_indexed_dataset.py]: Will this redownload the file on global rank 0?

File: megatron/core/datasets/s3_indexed_dataset.py [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+
+    if torch.distributed.is_initialized():
+        torch.distributed.barrier()
+
+    # If the `local_path` still does not exist, then we assume
+    # each rank is saving to a separate location.
+    if not os.path.exists(local_path):
+        _download_file(s3_client, s3_path, local_path)
```
Comment (jkamalu) [inline megatron/core/datasets/s3_indexed_dataset.py]: Will this redownload on all local rank 0s?

File: megatron/core/datasets/s3_indexed_dataset.py [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+
+    if torch.distributed.is_initialized():
+        torch.distributed.barrier()
+
+    assert os.path.exists(local_path)
+
+
+class S3IndexedDataset(torch.utils.data.Dataset):
```
Comment (jkamalu) [inline megatron/core/datasets/s3_indexed_dataset.py]: Can we make this class inherit from MMapIndexedDataset, keeping the length, get item, and properties, but overloading the s3 specific functions?

setstate and getstate could even be inherited if state were presumed defined in init

File: tests/unit_tests/data/test_preprocess_data.py [hunk 10]
```diff
@@ -184,25 +209,70 @@ def gpt2_merge(odir):
...
+        pass
+
+
+def test_preprocess_data_gpt_s3(monkeypatch):
+    MOCK_S3_CLIENT = MockS3Client()
+
+    def mock_s3_client(*args, **kwargs):
+        return MOCK_S3_CLIENT
+
+    monkeypatch.setattr("boto3.client", mock_s3_client)
```
Comment (jkamalu) [inline tests/unit_tests/data/test_preprocess_data.py:271]: Why would the following not work?

```
monkeypatch.setattr("boto3.client", MockS3Client)
```

File: tests/unit_tests/data/test_preprocess_data.py [hunk 11]
```diff
@@ -100,8 +104,27 @@ def tokens_to_string(toks):
...
+    if indexed_dataset_type == S3IndexedDataset:
+        # Copy all the files in `temp_dir` to S3.
+        s3_client = boto3.client("s3")
+        bucket_name = "test-bucket"
+        for path in glob.glob(os.path.join(temp_dir, "**/**")):
+            assert path.startswith("/")
+            s3_client.upload_file(path, bucket_name, path[1:])
+        assert path_to_data.startswith("/")
+        prefix_for_path_prefix = "s3://" + bucket_name
+        IndexedDataset = S3IndexedDataset
```
Comment (jkamalu) [inline tests/unit_tests/data/test_preprocess_data.py]: What purpose does this line serve?

File: tests/unit_tests/data/test_preprocess_data.py [hunk 12]
```diff
@@ -67,7 +71,7 @@ def merge_datasets(idir):
     merge_main()
 
 
-def do_test_preprocess_data(temp_dir, extra_args=[]):
+def do_test_preprocess_data(temp_dir, extra_args=[], indexed_dataset_type=MMapIndexedDataset):
```
Comment (jkamalu) [inline tests/unit_tests/data/test_preprocess_data.py]: No need for a default, let's make it a positional arg

**Review by jkamalu — COMMENTED** [review]
Left some comments. Thanks a bunch for the MR! Would like to see some consolidation in S3IndexedDataset before being merged.

File: tests/unit_tests/data/test_preprocess_data.py (hunk 10 above)
Comment (jrocmar) [reply tests/unit_tests/data/test_preprocess_data.py:271]: It doesn't work, because it creates a new MockS3Client each time boto3.client is called. Instead, I want the state (i.e., `self._data`) to be maintained across calls to boto3.client, because it reflects the state of the mock S3.

File: megatron/core/datasets/s3_indexed_dataset.py (hunk 1 above)
Comment (jrocmar) [reply megatron/core/datasets/s3_indexed_dataset.py]: Good point- I changed it to raise the exception if the error code is anything besides a 404.

File: megatron/core/datasets/s3_indexed_dataset.py (hunk 2 above)
Comment (jrocmar) [reply megatron/core/datasets/s3_indexed_dataset.py]: This logic is now moved into _S3BinReader. I changed it to use the parse_s3_path method.

File: megatron/core/datasets/s3_indexed_dataset.py (hunk 3 above)
Comment (jrocmar) [reply megatron/core/datasets/s3_indexed_dataset.py]: This logic is moved to _S3BinReader. Added arg and return typing for all functions.

File: megatron/core/datasets/s3_indexed_dataset.py (hunk 4 above)
Comment (jrocmar) [reply megatron/core/datasets/s3_indexed_dataset.py]: Done.

File: megatron/core/datasets/s3_indexed_dataset.py (hunk 5 above)
Comment (jrocmar) [reply megatron/core/datasets/s3_indexed_dataset.py]: Done.

File: megatron/core/datasets/s3_indexed_dataset.py (hunk 8 above)
Comment (jrocmar) [reply megatron/core/datasets/s3_indexed_dataset.py]: No, because `local_path` will already exist on local rank 0s.

File: tests/unit_tests/data/test_preprocess_data.py (hunk 11 above)
Comment (jrocmar) [reply tests/unit_tests/data/test_preprocess_data.py]: Removed after the refactor.

File: tests/unit_tests/data/test_preprocess_data.py (hunk 12 above)
Comment (jrocmar) [reply tests/unit_tests/data/test_preprocess_data.py]: This argument is now a boolean called `s3` that defaults to False. Do you still want to make the new argument a positional one?

File: megatron/core/datasets/s3_indexed_dataset.py (hunk 9 above)
Comment (jrocmar) [reply megatron/core/datasets/s3_indexed_dataset.py]: After some consideration and after seeing this [refactor](https://github.com/NVIDIA/Megatron-LM/commit/23f3f55b894f6a290590cc49c04b53d9540a7e59), it seemed like the best way to integrate S3 data loading into IndexedDataset was to introduce an abstract _BinReader class and concrete _MMapBinReader, _FileBinReader and _S3BinReader subclasses, but let me know your thoughts on the design. I'm open to feedback.

**Comment by jrocmar** [comment]
> Left some comments. Thanks a bunch for the MR! Would like to see some consolidation in S3IndexedDataset before being merged.

@jkamalu Thank you for the comments! Based on those comments and after seeing this [refactor](https://github.com/NVIDIA/Megatron-LM/commit/23f3f55b894f6a290590cc49c04b53d9540a7e59), I refactored S3 data loading to integrate it into IndexedDataset and address your feedback.

**Comment by jrocmar** [comment]
@jkamalu The _S3BinReader will have poor performance when using a global random shuffle over samples (which is what GPTDataset currently does). I need to either implement "block shuffling" in GPTDataset as described in the "Example" section [here](https://github.com/NVIDIA/NeMo/pull/8447) (that section also describes why _S3BinReader will have poor performance) or I need to add an option to disable shuffling in GPTDataset (the user then has to be responsible for preshuffling their data). I'm inclined to just add the option to disable shuffling to start, because it's simpler. What do you think?

**Comment by jrocmar** [comment]
> @jkamalu The _S3BinReader will have poor performance when using a global random shuffle over samples (which is what GPTDataset currently does). I need to either implement "block shuffling" in GPTDataset as described in the "Example" section [here](https://github.com/NVIDIA/NeMo/pull/8447) (that section also describes why _S3BinReader will have poor performance) or I need to add an option to disable shuffling in GPTDataset (the user then has to be responsible for preshuffling their data). I'm inclined to just add the option to disable shuffling to start, because it's simpler. What do you think?

Moving the "Example" section from the old NeMo PR into this comment.

In NeMo, a sample consists of `seq_length` tokens. For simplicity, suppose each token is 1 byte and `seq_length` is 100.

Each sample then takes 100 bytes.

Suppose we have a dataset with 12 samples.

Sample index 0 is stored in bytes [0, 100), sample index 1 is stored in bytes [100, 200), ..., and sample index 11 is stored in bytes [1100, 1200).

Currently, NeMo takes the list of sample indices:

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

And produces a `shuffle_idx`, which is just a permutation of those sample indices like:

[11, 3, 0, 6, 1, 9, 10, 2, 5, 7, 4, 8]

The `shuffle_idx` determines the order in which NeMo processes samples.

We could have the `IndexedDataset` just grab the bytes for a sample at a time. The first request would be for the bytes [1100, 1200), the second request would be for the bytes [300, 400), the third request would be for the bytes [0, 100) and so on in the order determined by `shuffle_idx`. That works, but it's slow, because you're making one request for each sample.

Let's try to introduce an in-memory cache. In particular, suppose the IndexedDataset does this:

* If the requested bytes range [`start`, `end`) is in the cache, then extract the requested bytes range from the cache.
* Otherwise, first refresh the cache by downloading the bytes range [`start`, `start` + `cache_nbytes`) and then extract the requested bytes range from the cache.

Suppose the `cache_nbytes` is 400. The first request would be for the bytes [1100, 1200). The cache is initially empty, so we refresh the cache by downloading the bytes [1100, 1500) and then extract the requested bytes range from the cache. The second request would be for the bytes [300, 400). Those bytes are not in the cache, so we refresh the cache by downloading the bytes [300, 700) and then extract the requested bytes range from that cache. And so on.

We actually made the problem worse. For most samples, we have to refresh the cache, so we have not reduced the number of requests much. We've just made the requests have to download a larger number of bytes. The issue is that the bytes needed for a sample index are probably not next to the bytes needed for the previous sample index.

To use the cache effectively, we have to introduce some correlation in the shuffle. In particular, we divide the original list of sample indices into blocks like:

* [0, 1, 2, 3]
* [4, 5, 6, 7]
* [8, 9, 10, 11]

We then shuffle within the blocks like:

* [3, 0, 2, 1]
* [4, 6, 5, 7]
* [11, 10, 8, 9]

We then shuffle the order of the blocks like:

* [11, 10, 8, 9]
* [4, 6, 5, 7]
* [3, 0, 2, 1]

And we construct the block-shuffled `shuffle_idx` like:

[11, 10, 8, 9, 4, 6, 5, 7, 3, 0, 2, 1]

We also have to change which bytes we download on a cache miss. In particular, we download the bytes [`cache_start`, `cache_start` + `cache_nbytes`), where cache_start is (`start`//`cache_nbytes`) * `cache_nbytes`.

The first request would be for the bytes [1100, 1200). The cache is initially empty, so we refresh the cache by downloading the bytes [800, 1200) and then extract the requested bytes range from that cache. The second request would be for the bytes [1000, 1100). We extract those bytes from the cache. The third request would be for the bytes [800, 1200). We extract those bytes from the cache. And so on. In this way, we only have to refresh cache at the start of each new block.

**Comment by ryxli** [comment]
Merged in https://github.com/NVIDIA/Megatron-LM/commit/a30a28dbe9063e8456ddc2f5ee1d26ede8589f63

Can mark as closed, thanks
